### ISSUE TITLE
`no-unused-expressions` should not allow unassigned/unreturned pure function calls

### ISSUE THREAD

**Issue by fregante** [eslint/eslint#14581]
**What rule do you want to change?**

`no-unused-expressions`

**Does this change cause the rule to produce more or fewer warnings?**

More

**How will the change be implemented? (New option, new default behavior, etc.)?**

New default behavior

**Please provide some example code that this change will affect:**

```js
let storage = [];
function store(...data) {
	storage.concat(bar);
}

store(1, 2, 3);
```

One might mistakenly think that `.concat` is one of the Array methods that mutate `this` or just forget to prepend `storage = `

This function will do nothing and actually the whole code will probably be dropped by a decent minifier.

The problem with implementing this would probably be detecting which function calls are side-effect-free, but it could at least start from some methods from the standard library.

**What does the rule currently do for this code?**

Nothing

**What will the rule do after it's changed?**

It should warn the user that the statement is a no-op, like it does for `window;`

**Comment by nzakas** [comment]
This isn’t a case we can handle. You’re asking us to know that storage is an array and that storage.concat() is a method on an array that doesn’t have side effects. This is a use case for a type checker, not for ESLint.
